README
LFS Book README

В этом документе содержится информация о том, как выполнить преобразование книги из XML формата в другой формат (например HTML, PDF, PS и TXT). Для начала, если вы ещё этого не сделали, пожалуйста, откройте и прочитайте информацию, представленную в файле INSTALL, чтобы установить необходимое программное обеспечение. Если вы уже выполнили все шаги в файле INSTALL, ознакомьтесь с примерами того, как выполнить преобразование файлов в другие форматы.

Во всех примерах, указание аргумента REV=systemd необходимо, если требуется сборка книги редакции systemd

XML в XHTML:
-------------
make BASEDIR=/path/to/output/location


XML в одностраничный XHTML (nochunks):
------------------------------------
make BASEDIR=/path/to/output/location nochunks


XML в TXT
----------
Follow the instructions for nochunks and then run:
lynx -dump /path/to/nochunks >/path/to/output


XML в PDF:
-----------
make BASEDIR=/path/to/output/location pdf




bootscripts/ChangeLog
2015-06-24 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Unmount network file systems before bringing down the network
   * Spell out some mount/umount options

2017-03-25 DJ Lucas <dj@linuxfromscratch.org>
   * Bash-4.4 changes no longer allow continue in function to pass to calling
     loop.

2015-02-22 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Cosmetic changes to network scripts' output

2015-02-08 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Modify ipv4-static to support ethernet labels
   * Output a newline if an interface is skipped during boot

2014-12-24 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Clarify rationale for sysklog run levels

2014-11-10 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Make /run/lock have 1777 permissions

2014-07-06 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Remove obsolete reference to /proc/sys/kernel/hotplug in udev boot script

2014-07-02 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Remove obsolete check for /proc/ksyms in modules boot script

2014-06-29 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Remove systemd references and files

2014-03-31 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Changes made to be consistent with common systemd functions

2013-08-21 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Delete duplicate entry in rc.site file.

2013-08-05 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Use STATIC_GATEWAY for ipv4-static-route purposes

2013-05-15 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Remove unneeded options and commands from mountfs and
     mountvirtfs when mtab is a symlink to /proc/self/mounts

2013-05-11 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Make sure the root fs is mounted read only in mountfs
     before halting

2013-01-24 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Remove local attribue of some variables for Bourne
     shell compatability
   * Specify /bin/echo in commands with -e switch for
     Bourne shell compatability

2012-10-13 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Remove unneeded sleep in killproc
   * Add option for verbose fsck
   * Flatten directory structure of tarball
   * Remove /fastboot or /forcecheck if set

2012-09-05 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Add udevadm trigger --action=change line to udev script
     to support initramfs better

2012-08-24 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Read /etc/sysconfig/console when needed in rc

2012-06-17 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Only mount /run in mountvirtfs if it is not already mounted
     (from an initrd).
   * Force the /dev/shm->/run/shm symlink
   * Skip making LVM virtual groups unavailable at shutdown
     because the / or swap partitions may be still in use.

2012-05-30 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Copy all entries in /lib/udev/devices to /dev in mounvirtfs

2012-05-24 Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Remove mkdir from mountvirtfs because fs is read only

2012-04-09 Bruce Dubbs <bdubbs@linuxfromscratch.org>

   * General update of network scripts ifup/ifdown and services
   ** Add if_true utility function to init_functions
   ** Change variable INTERFACES to COMPONENT_INTERFACES for bridge interface
   ** ifdown now only uses the first SERVICE
   ** Updated ifup.8 to reflect current usage
   ** ifup
   *** Move interface up logic to after service scripts.  Eliminates the
       need for the LINK_CHECK variable.
   *** Sets the GATEWAY if the variable is set
   *** Add MTU logic to ifup

2012-01-22  Matt Burgess <matthew@linuxfromscratch.org>

   * lfs/init.d/udev: Change udevd path following upstream changes to its
                      location

2011-12-23  Matt Burgess <matthew@linuxfromscratch.org>

   * lfs/init.d/mountvirtfs: Correct permissions on /run/shm

2011-11-28  DJ Lucas <dj@linuxfromscratch.org>

   * lfs/init.d/mountfs: Add '-t nosysfs' to umount command
   * lfs/init.d/mountvirtfs: remove unneeded spaces in output
   * lfs/init.d/rc: Correct double interactive prompt and set prompt defaults
   * lfs/init.d/setclock: Remove screen output from start case (udev)
   * lfs/init.d/udev{,_retry}:  Restore output to screen and log
   * lfs/sbin/ifup: Change "..." to ": " (not awaiting additional output)

2011-11-14  Bruce Dubbs <bdubbs@linuxfromscratch.org>

   * The following are from suggestions by DJ Lucas
   ** General cleanup of scripts
   ** Move consolelog function to rc
   ** Clean up boot logging functions
   ** Fix interactive prompt between run level S and target run level

2011-10-06  Bryan Kadzban <bryan@linuxfromscratch.org>

   * Add configuration for udev_retry, and remove --type=failed (which does
     not work with recent udev versions anyway, since no events can possibly
     trigger it).  Start with just the "rtc" subsystem.  BLFS should eventually
     add "sound" to this file, and perhaps others.

2011-09-18
   * Review and update of changes made in previous change.
   ** Rename /etc/sysconfig/init_params to /etc/sysconfig/rc.site.
   ** Move network services to /lib/services.
   ** Move init-functions to /lib/lsb.
   *** /lib/lsb is a symlink to /lib/services.
   * Create convenience symlink /etc/init.d->/etc/rc.d/init.d.
   * Add help and man pages to upup/ifdown.
   * Append /run/var/bootlog to /var/log/boot.log at the end of
     the boot sequence.
   * Add capabiltiy to step through the boot scripts at boot time
   * Optionally allow environment variables in sysconfig directory's
     console, network, and clock files to be placed in rc.site.
   * Add an optional FASTBOOT parameter to set /fastboot when rebooting.

2011-07-14  Bruce Dubbs <bdubbs@linuxfromscratch.org>
   * Major rewrite of all LFS scripts
   ** Move ifup/ifdown to /sbin
   ** Move functions and network services to /lib/boot
   ** Remove contrib directory
   ** Move all READ commands to functions and allow the user to specify
      HEADLESS=1 to skip READ commands

obfuscate.sh
#!/bin/bash

# obfuscate.sh
# obfuscate email addresses in XML/HTML
# Script written (and slight perl modification) by Archaic <archaic AT linuxfromscratch D0T org>
# Modified from "sed -i" to old style "sed -e" by Manuel Canales <manuel AT linuxfromscratch D0T org>
# to prevent hangs on very long files, like nonckunked books.
# Original Perl expression by Anderson Lizardo <lizardo AT linuxfromscratch D0T org>
# Released under the GNU General Public License
#
# This script currently only seeks out mailto: addresses. If those same
# addresses also appear in plaintext, we need to obfuscate those as well.
#
# This script was made for a very specific purpose so I was a bit lazy in
# writing the regex's.
#
# Please send comments, enhancements, etc. to the above address

#set -e  # Bail on all errors

# First, ensure that we are given a file to process
# if [ $# -lt 1 ]; then
#   echo -e "\nYou must provide an input file."
#   exit 1
# fi

# Nothing like a backup plan!
#cp "$1" "$1".bak

for i in `grep -o '"mailto:.*@.*"' ${1} |sed -e 's|^"mailto:||' -e 's|"$||'`; do
  link=`echo $i | perl -pe 's/[^\n]/"\\\&#".ord($&)."\;"/ge'`
  plaintext=`echo $i | sed -e 's|@| AT |' -e 's|\.| D0T |g'`
  cp ${1}{,.tmp}
  sed -e "s|mailto:$i|mailto:$link|" \
      -e "s|$i|$plaintext|" ${1}.tmp > ${1}
  rm ${1}.tmp
done

#rm $FILE.tmp
#exit 0


lfs-latest.php
#! /usr/bin/php
<?php
$dirs = array();
$vers = array();

date_default_timezone_set( "GMT" );
$date = date( "Y-m-d H:i:s" );

// Special cases
$exceptions = array();
//$exceptions[ 'gmp' ] = "UPDIR=/.*(gmp-\d[\d\.-]*\d).*/:DOWNDIR=";

$regex = array();
$regex[ 'bzip2'    ] = "/^.*current version is ([\d\.]+).*$/";
$regex[ 'check'    ] = "/^.*Check (\d[\d\.]+\d).*$/";
$regex[ 'intltool' ] = "/^.*Latest version is (\d[\d\.]+\d).*$/";
$regex[ 'less'     ] = "/^.*current released version is less-(\d+).*$/";
$regex[ 'mpfr'     ] = "/^mpfr-([\d\.]+)\.tar.*$/";
$regex[ 'systemd'  ] = "/^.*v([\d]+).*$/";
//$regex[ 'sysvinit' ] = "/^.*sysvinit-([\d\.]+)dsf\.tar.*$/";
$regex[ 'tzdata'   ] = "/^.*tzdata([\d]+[a-z]).*$/";
$regex[ 'xz'       ] = "/^.*xz-([\d\.]*\d).*$/";
$regex[ 'zlib'     ] = "/^.*zlib ([\d\.]*\d).*$/";

function find_max( $lines, $regex_match, $regex_replace )
{
  $a = array();
  if ( ! is_array( $lines ) ) return -1;

  foreach ( $lines as $line )
  {
     if ( ! preg_match( $regex_match, $line ) ) continue;

     // Isolate the version and put in an array
     $slice = preg_replace( $regex_replace, "$1", $line );
     if ( $slice == $line ) continue;

     array_push( $a, $slice );
  }

  // SORT_NATURAL requires php-5.4.0 or later
  rsort( $a, SORT_NATURAL );  // Max version is at the top
  return ( isset( $a[0] ) ) ? $a[0] : -2;
}

function find_even_max( $lines, $regex_match, $regex_replace )
{
  $a = array();
  foreach ( $lines as $line )
  {
     if ( ! preg_match( $regex_match, $line ) ) continue;

     // Isolate the version and put in an array
     $slice = preg_replace( $regex_replace, "$1", $line );

     if ( "x$slice" == "x$line" ) continue;

     // Skip odd numbered minor versions and minors > 80
     list( $major, $minor, $rest ) = explode( ".", $slice . ".0" );
     if ( $minor % 2 == 1  ) continue;
     if ( $minor     >  80 ) continue;
     array_push( $a, $slice );
  }

  rsort( $a, SORT_NATURAL );  // Max version is at the top
  return ( isset( $a[0] ) ) ? $a[0] : -2;
}

function http_get_file( $url )
{
  if ( ! preg_match( "/sourceforge/", $url ) )
  {
    exec( "curl --location --silent --max-time 30 $url", $dir );

    $s   = implode( "\n", $dir );
    $dir = strip_tags( $s );
    return explode( "\n", $dir );
  }
  else
  {
    exec( "links -dump $url 2>/dev/null", $lines );
    return $lines;
  }
}

function max_parent( $dirpath, $prefix )
{
  // First, remove a directory
  $dirpath  = rtrim  ( $dirpath, "/" );    // Trim any trailing slash
  $position = strrpos( $dirpath, "/" );
  $dirpath  = substr ( $dirpath, 0, $position );

  $lines = http_get_file( $dirpath );

  $regex_match   = "#${prefix}[\d\.]+/#";
  $regex_replace = "#^.*(${prefix}[\d\.]+)/.*$#";
  $max           = find_max( $lines, $regex_match, $regex_replace );

  return "$dirpath/$max";
}

function get_packages( $package, $dirpath )
{
  global $exceptions;
  global $regex;

//if ( $package != "expat" ) return 0; // Debug

if ( $package == "check"      ) $dirpath = "https://github.com/libcheck/check/releases";
if ( $package == "e2fsprogs"  ) $dirpath = "http://sourceforge.net/projects/e2fsprogs/files/e2fsprogs";
if ( $package == "expat"      ) $dirpath = "http://sourceforge.net/projects/expat/files";
if ( $package == "expect"     ) $dirpath = "http://sourceforge.net/projects/expect/files";
if ( $package == "file"       ) $dirpath = "https://github.com/file/file/releases";
if ( $package == "flex"       ) $dirpath = "https://github.com/westes/flex/releases";
if ( $package == "gcc"        ) $dirpath = max_parent( $dirpath, "gcc-" );
if ( $package == "intltool"   ) $dirpath = "https://launchpad.net/intltool/trunk";
if ( $package == "meson"      ) $dirpath = "https://github.com/mesonbuild/meson/releases";
if ( $package == "mpc"        ) $dirpath = "http://www.multiprecision.org/mpc/download.html";
if ( $package == "mpfr"       ) $dirpath = "http://mpfr.loria.fr/mpfr-current";
if ( $package == "ninja"      ) $dirpath = "https://github.com/ninja-build/ninja/releases";
if ( $package == "procps-ng"  ) $dirpath = "http://sourceforge.net/projects/procps-ng/files";
if ( $package == "psmisc"     ) $dirpath = "http://sourceforge.net/projects/$package/files";
if ( $package == "shadow"     ) $dirpath = "https://github.com/shadow-maint/shadow/releases";
if ( $package == "systemd"    ) $dirpath = "https://github.com/systemd/systemd/releases";
if ( $package == "tcl"        ) $dirpath = "http://sourceforge.net/projects/tcl/files";
if ( $package == "util-linux" ) $dirpath = max_parent( $dirpath, "v." );
if ( $package == "vim"        ) $dirpath = "ftp://ftp.vim.org/pub/vim/unix";

  // Check for ftp
  if ( preg_match( "/^ftp/", $dirpath ) )
  {
    $dirpath  = substr( $dirpath, 6 );           // Remove ftp://
    $dirpath  = rtrim ( $dirpath, "/" );         // Trim any trailing slash
    $position = strpos( $dirpath, "/" );         // Divide at first slash
    $server   = substr( $dirpath, 0, $position );
    $path     = substr( $dirpath, $position );

    $conn = ftp_connect( $server );
    ftp_login( $conn, "anonymous", "" );

    // See if we need special handling
    if ( isset( $exceptions[ $package ] ) )
    {
       $specials = explode( ":", $exceptions[ $package ] );

       foreach ( $specials as $i )
       {
          list( $op, $regexp ) = explode( "=", $i );

          switch ($op)
          {
            case "UPDIR":
              // Remove last dir from $path
              $position = strrpos( $path, "/" );
              $path = substr( $path, 0, $position );

              // Get dir listing
              $lines = ftp_rawlist ($conn, $path);
              $max   = find_max( $lines, $regexp, $regexp );
              break;

            case "DOWNDIR":
              // Append found directory
              $path .= "/$max";
              break;

            default:
              echo "Error in specials array for $package\n";
              return -5;
              break;
          }
       }
    }

    $lines = ftp_rawlist ($conn, $path);
    ftp_close( $conn );
  }
  else // http
  {
     // Customize http directories as needed
     if ( $package == "tzdata" )
     {
        // Remove two directories
        $dirpath  = rtrim  ( $dirpath, "/" );    // Trim any trailing slash
        $position = strrpos( $dirpath, "/" );
        $dirpath  = substr ( $dirpath, 0, $position );
        $position = strrpos( $dirpath, "/" );
        $dirpath  = substr ( $dirpath, 0, $position );
     }

     if ( $package == "bzip2" )
     {
        // Remove one directory
        $dirpath  = rtrim  ( $dirpath, "/" );    // Trim any trailing slash
        $position = strrpos( $dirpath, "/" );
        $dirpath  = substr ( $dirpath, 0, $position );
     }

     $lines = http_get_file( $dirpath );
     if ( ! is_array( $lines ) ) return -6;
  } // End fetch
//print_r($lines);
  if ( isset( $regex[ $package ] ) )
  {
     // Custom search for latest package name
     foreach ( $lines as $l )
     {
        $ver = preg_replace( $regex[ $package ], "$1", $l );
        if ( $ver == $l ) continue;
        return $ver;  // Return first match of regex
     }

     return -7;  // This is an error
  }

  if ( $package == "perl" )  // Custom for perl
  {
     $tmp = array();

     foreach ( $lines as $l )
     {
        if ( preg_match( "/sperl/", $l ) ) continue; // Don't want this
        $ver = preg_replace( "/^.*perl-([\d\.]+\d)\.tar.*$/", "$1", $l );
        if ( $ver == $l ) continue;
        list( $s1, $s2, $rest ) = explode( ".", $ver );
        if ( $s2 % 2 == 1 ) continue; // Remove odd minor versions
        array_push( $tmp, $l );
     }

     $lines = $tmp;
  }

  if ( $package == "attr" ||
       $package == "acl"  )
  {
     return find_max( $lines, "/$package/", "/^.*$package-([\d\.-]*\d)\.src.*$/" );
  }

  if ( $package == "e2fsprogs" )
     return find_max( $lines, "/v\d/", "/^.*v(\d[\d\.]+\d).*$/" );

  if ( $package == "expect" )
     return find_max( $lines, "/expect/", "/^.*expect(\d[\d\.]+\d).tar.*$/" );

  if ( $package == "mpc" )
     return find_max( $lines, "/Version/", "/^.*Version (\d[\d\.]+\d)\\s*$/" );

  if ( $package == "XML-Parser" )
  {
     $max = find_max( $lines, "/$package/", "/^.*$package-([\d\._]*\d).tar.*$/" );
     # 2.44_01 is a developer release
     if ( $max == "2.44_01" ) { return "2.44"; }
     return $max;
  }

  if ( $package == "tcl" )
     return find_max( $lines, "/tcl/", "/^.*tcl(\d[\d\.]*\d)-src.*$/" );

  if ( $package == "ninja" )
     return find_max( $lines, "/v\d/", "/^.*v(\d[\d\.]*\d).*$/" );

  if ( $package == "gmp" )
     return find_max( $lines, "/$package/", "/^.*$package-([\d\._]*\d[a-z]?).tar.*$/" );

  if ( $package == "dbus" )
     return find_even_max( $lines, "/$package/", "/^.*$package-([\d\.]+).tar.*$/" );

  if ( $package == "file" )
  {
     $max = find_max( $lines, "/FILE5/", "/^.*FILE(5_\d+)*$/" );
     return str_replace( "_", ".", $max );
  }

  if ( $package == "grub" )
     return find_max( $lines, "/grub/", "/^.*grub-(\d\..*).tar.xz.*$/" );

  if ( $package == "openssl" )
     return find_max( $lines, "/openssl/", "/^.*openssl-([\d\.p]*\d.?).tar.*$/" );

  // Most packages are in the form $package-n.n.n
  // Occasionally there are dashes (e.g. 201-1)
  return find_max( $lines, "/$package/", "/^.*$package-([\d\.-]*\d)\.tar.*$/" );
}

function get_current()
{
   global $dirs;
   global $vers;

   // Fetech from svn and get wget-list
   $current = array();
   $lfssvn = "svn://svn.linuxfromscratch.org/LFS/trunk";

   $tmpdir = exec( "mktemp -d /tmp/lfscheck.XXXXXX" );
   $cdir   = getcwd();
   chdir( $tmpdir );
   exec ( "svn --quiet export $lfssvn LFS" );
   chdir( $cdir );

   $PAGE       = "$tmpdir/LFS/BOOK/chapter03/chapter03.xml";
   $STYLESHEET = "$tmpdir/LFS/BOOK/stylesheets/wget-list.xsl";

   exec( "xsltproc --xinclude --nonet $STYLESHEET $PAGE", $current );
   exec( "rm -rf $tmpdir" );

   foreach ( $current as $line )
   {
      $file = basename( $line ) . "\n";
      if ( preg_match( "/patch$/", $file ) ) { continue; } // Skip patches

      $file = preg_replace( "/bz2/", '', $file ); // The 2 confusses the regex

      $file        = rtrim( $file );
      $pkg_pattern = "/(\D*).*/";
      //$pattern     = "/\D*(\d.*\d)\D*/";
      $pattern     = "/\D*(\d.*\d)\D*/";

      if ( preg_match( "/e2fsprogs/", $file ) )
      {
        $pattern = "/e2\D*(\d.*\d)\D*/";
        $pkg_pattern = "/(e2\D*).*/";
      }

      else if ( preg_match( "/tzdata/", $file ) )
      {
        $pattern = "/\D*(\d.*[a-z])\.tar\D*/";
      }

      else if ( preg_match( "/openssl/", $file ) )
      {
        $pattern = "/\D*(\d.*\d.*).tar.*$/";
      }

      else if ( preg_match( "/gmp/", $file ) )
      {
        $pattern = "/\D*(\d.*[a-z]*)\.tar\D*/";
      }

      else if ( preg_match( "/systemd-man-pages/", $file ) ) continue;
      else if ( preg_match( "/python/"         , $file ) ) continue;

      $version = preg_replace( $pattern, "$1", $file );   // Isolate version
      $version = preg_replace( "/^\d-/", "", $version );  // Remove leading #-

      // Touch up package names
      $pkg_name = preg_replace( $pkg_pattern, "$1", $file );
      $pkg_name = trim( $pkg_name, "-" );

      if ( preg_match( "/bzip|iproute/", $pkg_name ) ) { $pkg_name .= "2"; }
      if ( preg_match( "/^m$/"         , $pkg_name ) ) { $pkg_name .= "4"; }
      if ( preg_match( "/shadow/"      , $pkg_name ) ) { $pkg_name  = "shadow"; }

      $dirs[ $pkg_name ] = dirname( $line );
      $vers[ $pkg_name ] = $version;
   }
}

function mail_to_lfs()
{
   global $date;
   global $vers;
   global $dirs;

   //$to      = "bruce.dubbs@gmail.com";
   $to      = "lfs-book@lists.linuxfromscratch.org";
   $from    = "bdubbs@linuxfromscratch.org";
   $subject = "LFS Package Currency Check - $date GMT";
   $headers = "From: bdubbs@anduin.linuxfromscratch.org";

   $message = "Package         LFS      Upstream  Flag\n\n";

   foreach ( $dirs as $pkg => $dir )
   {
      //if ( $pkg != "gmp" ) continue;  //debug
      $v = get_packages( $pkg, $dir );

      $flag = ( $vers[ $pkg ] != $v ) ? "*" : "";

      // Pad for output
      $pad = "                ";
      $p   = substr( $pkg          . $pad, 0, 15 );
      $l   = substr( $vers[ $pkg ] . $pad, 0, 10 );
      $c   = substr( $v            . $pad, 0, 10 );

      $message .= "$p $l $c $flag\n";
   }

   exec ( "echo '$message' | mailx -r $from -s '$subject' $to" );
   //echo $message;
}

function html()
{

   global $date;
   global $vers;
   global $dirs;

   echo "<!DOCTYPE html PUBLIC '-//W3C//DTD XHTML 1.0 Strict//EN'
                      'http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd'>
<html xmlns='http://www.w3.org/1999/xhtml' xml:lang='en' lang='en'>
<head>
<title>LFS Package Currency Check - $date</title>
<style type='text/css'>
h1, h2 {
   text-align      : center;
}

table {
   border-width    : 1px;
   border-spacing  : 0px;
   border-style    : outset;
   border-color    : gray;
   border-collapse : separate;
   background-color: white;
   margin          : 0px auto;
}

table th {
   border-width    : 1px;
   padding         : 2px;
   border-style    : inset;
   border-color    : gray;
   background-color: white;
}

table td {
   border-width    : 1px;
   padding         : 2px;
   border-style    : inset;
   border-color    : gray;
   background-color: white;
}
</style>

</head>
<body>
<h1>LFS Package Currency Check</h1>
<h2>As of $date GMT</h1>

<table>
<tr><th>LFS Package</th> <th>LFS Version</th> <th>Latest</th> <th>Flag</th></tr>\n";

   // Get the latest version of each package
   foreach ( $dirs as $pkg => $dir )
   {
      $v    = get_packages( $pkg, $dir );
      $flag = ( $vers[ $pkg ] != $v ) ? "*" : "";
      echo "<tr><td>$pkg</td> <td>${vers[ $pkg ]}</td> <td>$v</td> <td>$flag</td></tr>\n";
   }

   echo "</table>
</body>
</html>\n";

}

get_current();  // Get what is in the book
mail_to_lfs();
//html();  // Write html output
?>
